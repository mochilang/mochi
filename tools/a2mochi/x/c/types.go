package c

type function struct {
	name      string
	ret       string
	params    []param
	body      []string
	startLine int
	endLine   int
	source    string
	signature string
}

type param struct {
	name string
	typ  string
}
