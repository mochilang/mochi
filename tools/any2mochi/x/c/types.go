package c

// function describes a parsed C function.
// Additional location information is useful for better diagnostics
// and potential future features.
type function struct {
	name      string
	ret       string
	params    []param
	body      []string
	startLine int // 1-indexed line of the function definition
	endLine   int // 1-indexed line of the closing brace
}

type param struct {
	name string
	typ  string
}
