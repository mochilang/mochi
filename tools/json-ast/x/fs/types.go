package fs

type Var struct {
	Name    string `json:"name"`
	Expr    string `json:"expr"`
	Mutable bool   `json:"mutable"`
	Type    string `json:"type"`
	Line    int    `json:"line"`
	Raw     string `json:"raw"`
}

type Assign struct {
	Name  string `json:"name"`
	Index string `json:"index"`
	Expr  string `json:"expr"`
	Line  int    `json:"line"`
	Raw   string `json:"raw"`
}

type ForIn struct {
	Var  string `json:"var"`
	Expr string `json:"expr"`
	Body []Stmt `json:"body"`
	Line int    `json:"line"`
	Raw  string `json:"raw"`
}

type PrintStmt struct {
	Expr string `json:"expr"`
	Line int    `json:"line"`
	Raw  string `json:"raw"`
}

type Stmt interface{}
