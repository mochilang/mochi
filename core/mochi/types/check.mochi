package types

import "core/mochi/parser/parser.mochi" as mp

type Result {
  typ: string
  errs: list<string>
}

// Simple type checker for the toy AST used by parser.mochi.
// It supports integers, unary negation and the operators +, -, *, / with
// variable declarations.

fun exprType(node: any, env: map<string,string>): Result {
  if node.kind == "int" {
    return Result{ typ: "int", errs: [] }
  }
  if node.kind == "ident" {
    let val = env[node.value]
    if val {
      return Result{ typ: val, errs: [] }
    }
    return Result{ typ: "any", errs: ["unknown variable: " + node.value] }
  }
  if node.kind == "unary" {
    let res = exprType(node.children[0], env)
    if res.typ == "int" {
      return Result{ typ: "int", errs: res.errs }
    }
    return Result{ typ: "any", errs: res.errs + ["type mismatch around unary " + node.value] }
  }
  if node.kind == "binary" {
    let op = node.value
    let resL = exprType(node.children[0], env)
    let resR = exprType(node.children[1], env)
    var errs = resL.errs + resR.errs
    if resL.typ == "int" && resR.typ == "int" {
      return Result{ typ: "int", errs: errs }
    }
    errs = errs + ["type mismatch around operator " + op]
    return Result{ typ: "any", errs: errs }
  }
  return Result{ typ: "any", errs: ["unsupported expr: " + node.kind] }
}

fun checkProg(node: any): list<string> {
  var env: map<string,string> = {}
  var errs: list<string> = []
  for stmt in node.children {
    if stmt.kind == "let" || stmt.kind == "var" {
      let res = exprType(stmt.children[0], env)
      if count(res.errs) > 0 { errs = errs + res.errs }
      env[stmt.value] = res.typ
    } else if stmt.kind == "assign" {
      let res = exprType(stmt.children[0], env)
      if count(res.errs) > 0 { errs = errs + res.errs }
      if stmt.value in env {
        let cur = env[stmt.value]
        if cur != res.typ {
          errs = errs + ["type mismatch assigning to " + stmt.value]
        }
        env[stmt.value] = res.typ
      } else {
        errs = errs + ["unknown variable: " + stmt.value]
      }
    }
  }
  return errs
}

export fun checkString(src: string): string {
  let ast = mp.parseNode(src)
  let errs = checkProg(ast)
  if count(errs) == 0 { return "✅ Type Check Passed" }
  var out = "❌ Type Check Failed\n"
  var i = 1
  for e in errs {
    out = out + "  " + str(i) + ". " + e + "\n"
    i = i + 1
  }
  return out
}
