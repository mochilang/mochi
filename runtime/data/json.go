package data

import (
	"encoding/json"
	"fmt"
	"io"
)

// LoadJSON reads a JSON file containing either an object or array of objects.
func LoadJSON(path string) ([]map[string]any, error) {
	r, closeFn, err := openReader(path)
	if err != nil {
		return nil, err
	}
	defer closeFn()
	return LoadJSONReader(r)
}

// LoadJSONReader reads JSON data from r and returns rows.
func LoadJSONReader(r io.Reader) ([]map[string]any, error) {
	dec := json.NewDecoder(r)
	var data any
	if err := dec.Decode(&data); err != nil {
		return nil, err
	}
	switch v := data.(type) {
	case []any:
		out := make([]map[string]any, 0, len(v))
		for _, it := range v {
			m, ok := it.(map[string]any)
			if !ok {
				return nil, fmt.Errorf("json array element must be object")
			}
			out = append(out, m)
		}
		return out, nil
	case map[string]any:
		return []map[string]any{v}, nil
	default:
		return nil, fmt.Errorf("unsupported json value %T", data)
	}
}

// SaveJSON writes rows to a JSON file as an array or single object.
func SaveJSON(rows []map[string]any, path string) error {
	w, closeFn, err := openWriter(path)
	if err != nil {
		return err
	}
	defer closeFn()
	return SaveJSONWriter(rows, w)
}

// SaveJSONWriter writes rows to w in JSON format.
func SaveJSONWriter(rows []map[string]any, w io.Writer) error {
	enc := json.NewEncoder(w)
	if len(rows) == 1 {
		return enc.Encode(rows[0])
	}
	return enc.Encode(rows)
}
