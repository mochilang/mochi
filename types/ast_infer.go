package types

import "mochi/ast"

// NodeType returns the static type of the given AST node using env when
// available. It handles only a subset of node kinds sufficient for the
// COBOL backend.
func NodeType(n *ast.Node, env *Env) Type {
	if n == nil {
		return AnyType{}
	}
	switch n.Kind {
	case "int":
		return IntType{}
	case "float":
		return FloatType{}
	case "string":
		return StringType{}
	case "bool":
		return BoolType{}
	case "list":
		var elem Type = AnyType{}
		if len(n.Children) > 0 {
			elem = NodeType(n.Children[0], env)
		}
		return ListType{Elem: elem}
	case "selector":
		if env != nil {
			if t, err := env.GetVar(n.Value.(string)); err == nil {
				return t
			}
		}
		return AnyType{}
	case "call":
		if env != nil {
			if t, err := env.GetVar(n.Value.(string)); err == nil {
				if ft, ok := t.(FuncType); ok {
					return ft.Return
				}
			}
		}
		return AnyType{}
	case "binary":
		lt := NodeType(n.Children[0], env)
		rt := NodeType(n.Children[1], env)
		op := n.Value.(string)
		switch op {
		case "+", "-", "*", "/", "%":
			if isFloat(lt) || isFloat(rt) {
				return FloatType{}
			}
			if isInt(lt) && isInt(rt) {
				return IntType{}
			}
			if op == "+" && isString(lt) && isString(rt) {
				return StringType{}
			}
			if op == "+" && isList(lt) && isList(rt) {
				llt := lt.(ListType)
				rlt := rt.(ListType)
				if equalTypes(llt.Elem, rlt.Elem) {
					return llt
				}
			}
			return AnyType{}
		case "==", "!=", "<", "<=", ">", ">=", "&&", "||":
			return BoolType{}
		}
	case "group":
		return NodeType(n.Children[0], env)
	case "unary":
		return NodeType(n.Children[0], env)
	case "index":
		base := NodeType(n.Children[0], env)
		if lt, ok := base.(ListType); ok {
			if len(n.Children) > 1 && (n.Children[1].Kind == "start" || n.Children[1].Kind == "end" || len(n.Children) > 2) {
				return lt
			}
			return lt.Elem
		}
		if isString(base) {
			return StringType{}
		}
	case "if_expr":
		thenT := NodeType(n.Children[1], env)
		if len(n.Children) > 2 {
			elseT := NodeType(n.Children[2], env)
			if equalTypes(thenT, elseT) {
				return thenT
			}
			return AnyType{}
		}
		return thenT
	case "match":
		var t Type
		for _, cs := range n.Children[1:] {
			r := NodeType(cs.Children[1], env)
			if t == nil {
				t = r
				continue
			}
			if !equalTypes(t, r) {
				t = AnyType{}
			}
		}
		if t == nil {
			return AnyType{}
		}
		return t
	}
	return AnyType{}
}

// IsString reports whether n has static type string.
func IsString(n *ast.Node, env *Env) bool {
	_, ok := NodeType(n, env).(StringType)
	return ok
}

// IsFloat reports whether n has static type float.
func IsFloat(n *ast.Node, env *Env) bool {
	_, ok := NodeType(n, env).(FloatType)
	return ok
}

// IsList reports whether n has static list type.
func IsList(n *ast.Node, env *Env) bool {
	_, ok := NodeType(n, env).(ListType)
	return ok
}
