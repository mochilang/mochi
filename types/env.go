package types

import (
	"fmt"
	"io"
	"mochi/parser"
	"os"
)

// ModelSpec defines a named model configuration.
type ModelSpec struct {
	Provider string
	Name     string
	Params   map[string]any
}

// Env holds both type and value bindings for variables and functions.
type Env struct {
	parent *Env

	types   map[string]Type              // static types
	structs map[string]StructType        // user-defined struct types
	unions  map[string]UnionType         // user-defined union types
	streams map[string]StructType        // stream declarations
	agents  map[string]*parser.AgentDecl // agent declarations
	mut     map[string]bool              // mutability of variables
	values  map[string]any               // runtime values
	funcs   map[string]*parser.FunStmt   // function declarations
	models  map[string]ModelSpec         // model aliases

	output io.Writer // default: os.Stdout
	input  io.Reader // default: os.Stdin
}

// Types exposes the map of locally declared variable types.
func (e *Env) Types() map[string]Type { return e.types }

// NewEnv creates a new lexical scope environment.
func NewEnv(parent *Env) *Env {
	var out io.Writer = os.Stdout
	var in io.Reader = os.Stdin
	if parent != nil {
		out = parent.output
		in = parent.input
	}
	return &Env{
		parent:  parent,
		types:   make(map[string]Type),
		structs: make(map[string]StructType),
		unions:  make(map[string]UnionType),
		streams: make(map[string]StructType),
		agents:  make(map[string]*parser.AgentDecl),
		mut:     make(map[string]bool),
		values:  make(map[string]any),
		funcs:   make(map[string]*parser.FunStmt),
		models:  make(map[string]ModelSpec),
		output:  out,
		input:   in,
	}
}

// --- Type (Static) Binding ---

// SetStruct defines a user-defined struct type.
func (e *Env) SetStruct(name string, st StructType) {
	e.structs[name] = st
}

// GetStruct retrieves a struct type by name.
func (e *Env) GetStruct(name string) (StructType, bool) {
	if t, ok := e.structs[name]; ok {
		return t, true
	}
	if e.parent != nil {
		return e.parent.GetStruct(name)
	}
	return StructType{}, false
}

// Structs returns all struct types visible in the environment.
func (e *Env) Structs() map[string]StructType {
	out := map[string]StructType{}
	if e.parent != nil {
		for k, v := range e.parent.Structs() {
			out[k] = v
		}
	}
	for k, v := range e.structs {
		out[k] = v
	}
	return out
}

// SetUnion defines a user-defined union type.
func (e *Env) SetUnion(name string, ut UnionType) { e.unions[name] = ut }

// GetUnion retrieves a union type by name.
func (e *Env) GetUnion(name string) (UnionType, bool) {
	if u, ok := e.unions[name]; ok {
		return u, true
	}
	if e.parent != nil {
		return e.parent.GetUnion(name)
	}
	return UnionType{}, false
}

// SetStream defines a stream declaration.
func (e *Env) SetStream(name string, st StructType) { e.streams[name] = st }

// GetStream retrieves a stream by name.
func (e *Env) GetStream(name string) (StructType, bool) {
	if st, ok := e.streams[name]; ok {
		return st, true
	}
	if e.parent != nil {
		return e.parent.GetStream(name)
	}
	return StructType{}, false
}

// FindUnionByVariant returns the union type that contains the given variant name.
func (e *Env) FindUnionByVariant(variant string) (UnionType, bool) {
	for _, u := range e.unions {
		if _, ok := u.Variants[variant]; ok {
			return u, true
		}
	}
	if e.parent != nil {
		return e.parent.FindUnionByVariant(variant)
	}
	return UnionType{}, false
}

// SetModel defines a model alias.
func (e *Env) SetModel(name string, spec ModelSpec) { e.models[name] = spec }

// SetAgent defines an agent declaration.
func (e *Env) SetAgent(name string, decl *parser.AgentDecl) { e.agents[name] = decl }

// GetModel retrieves a model alias.
func (e *Env) GetModel(name string) (ModelSpec, bool) {
	if m, ok := e.models[name]; ok {
		return m, true
	}
	if e.parent != nil {
		return e.parent.GetModel(name)
	}
	return ModelSpec{}, false
}

// GetAgent retrieves an agent declaration.
func (e *Env) GetAgent(name string) (*parser.AgentDecl, bool) {
	if a, ok := e.agents[name]; ok {
		return a, true
	}
	if e.parent != nil {
		return e.parent.GetAgent(name)
	}
	return nil, false
}

// SetVar defines a variable's static type.
func (e *Env) SetVar(name string, typ Type, mutable bool) {
	e.types[name] = typ
	e.mut[name] = mutable
}

// SetVarDeep updates the type of an existing variable in the current
// environment or any parent scope. If the variable is not already defined,
// it is created in the current scope.
func (e *Env) SetVarDeep(name string, typ Type, mutable bool) {
	if _, ok := e.types[name]; ok || e.parent == nil {
		e.types[name] = typ
		e.mut[name] = mutable
		return
	}
	e.parent.SetVarDeep(name, typ, mutable)
}

// GetVar looks up a variable's static type.
func (e *Env) GetVar(name string) (Type, error) {
	if t, ok := e.types[name]; ok {
		return t, nil
	}
	if e.parent != nil {
		return e.parent.GetVar(name)
	}
	return nil, fmt.Errorf("undefined variable: %s", name)
}

func (e *Env) isMutable(name string) (bool, bool) {
	if m, ok := e.mut[name]; ok {
		return m, true
	}
	if e.parent != nil {
		return e.parent.isMutable(name)
	}
	return false, false
}

func (e *Env) IsMutable(name string) (bool, error) {
	if m, ok := e.isMutable(name); ok {
		return m, nil
	}
	return false, fmt.Errorf("undefined variable: %s", name)
}

// --- Value (Runtime) Binding ---

// SetValue sets a variable's runtime value.
func (e *Env) SetValue(name string, val any, mutable bool) {
	e.values[name] = val
	e.mut[name] = mutable
}

// UpdateValue modifies an existing variable's runtime value.
func (e *Env) UpdateValue(name string, val any) error {
	if _, ok := e.values[name]; ok {
		mutable, _ := e.isMutable(name)
		if !mutable {
			return fmt.Errorf("cannot assign to immutable variable: %s", name)
		}
		e.values[name] = val
		return nil
	}
	if e.parent != nil {
		return e.parent.UpdateValue(name, val)
	}
	return fmt.Errorf("variable not declared: %s", name)
}

// GetValue retrieves a runtime value.
func (e *Env) GetValue(name string) (any, error) {
	if val, ok := e.values[name]; ok {
		return val, nil
	}
	if e.parent != nil {
		return e.parent.GetValue(name)
	}
	return nil, fmt.Errorf("undefined variable: %s", name)
}

// --- Function Binding ---

// SetFunc binds a named function.
func (e *Env) SetFunc(name string, fn *parser.FunStmt) {
	e.funcs[name] = fn
}

// SetFuncType binds a function name to its static type.
func (e *Env) SetFuncType(name string, typ Type) {
	e.types[name] = typ
}

// GetFunc retrieves a function definition.
func (e *Env) GetFunc(name string) (*parser.FunStmt, bool) {
	if fn, ok := e.funcs[name]; ok {
		return fn, true
	}
	if e.parent != nil {
		return e.parent.GetFunc(name)
	}
	return nil, false
}

// Copy creates a shallow copy of the current environment with no parent.
// Useful for closures capturing current bindings.
func (e *Env) Copy() *Env {
	newEnv := &Env{
		parent:  nil, // flatten parent chain
		types:   make(map[string]Type, len(e.types)),
		mut:     make(map[string]bool, len(e.mut)),
		values:  make(map[string]any, len(e.values)),
		funcs:   make(map[string]*parser.FunStmt, len(e.funcs)),
		structs: make(map[string]StructType, len(e.structs)),
		unions:  make(map[string]UnionType, len(e.unions)),
		streams: make(map[string]StructType, len(e.streams)),
		agents:  make(map[string]*parser.AgentDecl, len(e.agents)),
		models:  make(map[string]ModelSpec, len(e.models)),
		output:  e.output,
		input:   e.input,
	}
	for k, v := range e.types {
		newEnv.types[k] = v
	}
	for k, v := range e.mut {
		newEnv.mut[k] = v
	}
	for k, v := range e.values {
		newEnv.values[k] = v
	}
	for k, v := range e.funcs {
		newEnv.funcs[k] = v
	}
	for k, v := range e.structs {
		newEnv.structs[k] = v
	}
	for k, v := range e.unions {
		newEnv.unions[k] = v
	}
	for k, v := range e.streams {
		newEnv.streams[k] = v
	}
	for k, v := range e.agents {
		newEnv.agents[k] = v
	}
	for k, v := range e.models {
		newEnv.models[k] = v
	}
	return newEnv
}

// --- Output Control ---

// SetWriter sets the output destination.
func (e *Env) SetWriter(w io.Writer) {
	e.output = w
}

// Writer returns the current output writer.
func (e *Env) Writer() io.Writer {
	return e.output
}

// SetReader sets the input source.
func (e *Env) SetReader(r io.Reader) { e.input = r }

// Reader returns the current input reader.
func (e *Env) Reader() io.Reader { return e.input }
