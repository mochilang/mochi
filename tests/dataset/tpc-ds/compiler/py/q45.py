# Generated by Mochi compiler v0.10.27 on 2025-07-17T17:50:14Z
from __future__ import annotations
import dataclasses
import json


@dataclasses.dataclass
class Auto1:
    ca_zip: str
    sum_ws_sales_price: float

    def __getitem__(self, key):
        return getattr(self, key)

    def __contains__(self, key):
        return hasattr(self, key)


@dataclasses.dataclass
class Customer:
    c_customer_sk: int
    c_current_addr_sk: int

    def __getitem__(self, key):
        return getattr(self, key)

    def __contains__(self, key):
        return hasattr(self, key)


@dataclasses.dataclass
class CustomerAddres:
    ca_address_sk: int
    ca_zip: str

    def __getitem__(self, key):
        return getattr(self, key)

    def __contains__(self, key):
        return hasattr(self, key)


@dataclasses.dataclass
class DateDim:
    d_date_sk: int
    d_qoy: int
    d_year: int

    def __getitem__(self, key):
        return getattr(self, key)

    def __contains__(self, key):
        return hasattr(self, key)


@dataclasses.dataclass
class Item:
    i_item_sk: int
    i_item_id: str

    def __getitem__(self, key):
        return getattr(self, key)

    def __contains__(self, key):
        return hasattr(self, key)


@dataclasses.dataclass
class WebSale:
    bill_customer_sk: int
    item_sk: int
    sold_date_sk: int
    sales_price: float

    def __getitem__(self, key):
        return getattr(self, key)

    def __contains__(self, key):
        return hasattr(self, key)


from typing import Any, TypeVar, Generic, Callable

T = TypeVar("T")
K = TypeVar("K")
UNDEFINED = object()


class _Group(Generic[K, T]):

    def __init__(self, key: K):
        self.key = key
        self.Items: list[T] = []
        self.items = self.Items

    def __iter__(self):
        return iter(self.Items)

    def __len__(self):
        return len(self.Items)


def _group_by(src: list[T], keyfn: Callable[[T], K]) -> list[_Group[K, T]]:
    groups: dict[str, _Group[K, T]] = {}
    order: list[str] = []
    for it in src:
        if isinstance(it, (list, tuple)):
            key = keyfn(*it)
        else:
            key = keyfn(it)
        if isinstance(key, dict):
            import types

            key = types.SimpleNamespace(**key)
        ks = str(key)
        g = groups.get(ks)
        if not g:
            g = _Group(key)
            groups[ks] = g
            order.append(ks)
        g.Items.append(it)
    return [groups[k] for k in order]


def _query(src, joins, opts):
    items = [[v] for v in src]
    for j in joins:
        joined = []
        if j.get("right") and j.get("left"):
            matched = [False] * len(j["items"])
            for left in items:
                m = False
                for ri, right in enumerate(j["items"]):
                    keep = True
                    if j.get("on"):
                        keep = j["on"](*left, right)
                    if not keep:
                        continue
                    m = True
                    matched[ri] = True
                    joined.append(left + [right])
                if not m:
                    joined.append(left + [None])
            for ri, right in enumerate(j["items"]):
                if not matched[ri]:
                    undef = [None] * (len(items[0]) if items else 0)
                    joined.append(undef + [right])
        elif j.get("right"):
            for right in j["items"]:
                m = False
                for left in items:
                    keep = True
                    if j.get("on"):
                        keep = j["on"](*left, right)
                    if not keep:
                        continue
                    m = True
                    joined.append(left + [right])
                if not m:
                    undef = [None] * (len(items[0]) if items else 0)
                    joined.append(undef + [right])
        else:
            for left in items:
                m = False
                for right in j["items"]:
                    keep = True
                    if j.get("on"):
                        keep = j["on"](*left, right)
                    if not keep:
                        continue
                    m = True
                    joined.append(left + [right])
                if j.get("left") and (not m):
                    joined.append(left + [None])
        items = joined
    if opts.get("where"):
        items = [r for r in items if opts["where"](*r)]
    if opts.get("sortKey"):

        def _key(it):
            k = opts["sortKey"](*it)
            if isinstance(k, (list, tuple, dict)):
                return str(k)
            return k

        items.sort(key=_key)
    if "skip" in opts:
        n = opts["skip"]
        if n < 0:
            n = 0
        items = items[n:] if n < len(items) else []
    if "take" in opts:
        n = opts["take"]
        if n < 0:
            n = 0
        items = items[:n] if n < len(items) else items
    res = []
    for r in items:
        res.append(opts["select"](*r))
    return res


def _sum(v):
    if hasattr(v, "Items"):
        v = v.Items
    if not isinstance(v, list):
        raise Exception("sum() expects list or group")
    s = 0.0
    for it in v:
        if it is None:
            continue
        if isinstance(it, (int, float)):
            s += float(it)
        else:
            raise Exception("sum() expects numbers")
    return s


def test_TPCDS_Q45_simplified():
    assert records == [
        Auto1(ca_zip="85669", sum_ws_sales_price=50.0),
        Auto1(ca_zip="99999", sum_ws_sales_price=30.0),
    ]


web_sales = [
    WebSale(bill_customer_sk=1, item_sk=1, sold_date_sk=1, sales_price=50.0),
    WebSale(bill_customer_sk=2, item_sk=2, sold_date_sk=1, sales_price=30.0),
]
customer = [
    Customer(c_customer_sk=1, c_current_addr_sk=1),
    Customer(c_customer_sk=2, c_current_addr_sk=2),
]
customer_address = [
    CustomerAddres(ca_address_sk=1, ca_zip="85669"),
    CustomerAddres(ca_address_sk=2, ca_zip="99999"),
]
item = [Item(i_item_sk=1, i_item_id="I1"), Item(i_item_sk=2, i_item_id="I2")]
date_dim = [DateDim(d_date_sk=1, d_qoy=1, d_year=2020)]
zip_list = [
    "85669",
    "86197",
    "88274",
    "83405",
    "86475",
    "85392",
    "85460",
    "80348",
    "81792",
]
item_ids = ["I2"]
qoy = 1
year = 2020


def _q0():
    _src = web_sales
    _rows = _query(
        _src,
        [
            {
                "items": customer,
                "on": lambda ws, c: ws.bill_customer_sk == c.c_customer_sk,
            },
            {
                "items": customer_address,
                "on": lambda ws, c, ca: c.c_current_addr_sk == ca.ca_address_sk,
            },
            {"items": item, "on": lambda ws, c, ca, i: ws.item_sk == i.i_item_sk},
            {
                "items": date_dim,
                "on": lambda ws, c, ca, i, d: ws.sold_date_sk == d.d_date_sk,
            },
        ],
        {
            "select": lambda ws, c, ca, i, d: (ws, c, ca, i, d),
            "where": lambda ws, c, ca, i, d: (
                (ca.ca_zip[0:5] in zip_list or i.i_item_id in item_ids)
                and d.d_qoy == qoy
            )
            and d.d_year == year,
        },
    )
    _groups = _group_by(_rows, lambda ws, c, ca, i, d: ca.ca_zip)
    _items1 = _groups
    return [
        Auto1(ca_zip=g.key, sum_ws_sales_price=_sum([x[0].sales_price for x in g]))
        for g in _items1
    ]


base = _q0()
records = base
print(json.dumps(records, separators=(",", ":"), default=lambda o: vars(o)))
test_TPCDS_Q45_simplified()
