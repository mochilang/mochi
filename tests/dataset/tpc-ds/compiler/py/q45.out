[{"ca_zip": "85669", "sum_ws_sales_price": 50.0}, {"ca_zip": "99999", "sum_ws_sales_price": 30.0}]
